pkcs11/pkcs11.go

```diff
@@ -1687,3 +1734,160 @@ func (s *Slot) generateECDSA(o keyOptions) (crypto.PrivateKey, error) {
 	}
 	return priv, nil
 }
+
+func (r *rsaPrivateKey) WithHash(hash crypto.Hash) (*rsaPrivateKey, error) {
+	r.hash = &hash
+	return r, nil
+}
+
+func (r *rsaPrivateKey) getHash() *crypto.Hash {
+	if r.hash != nil {
+		return r.hash
+	}
+	// Initialized to SHA256 if no hash specified
+	hash := crypto.SHA256
+	return &hash
+}
+
+func (r *rsaPrivateKey) WithPublicKeyHandle(o Object) *rsaPrivateKey {
+	r.pubH = &o
+	return r
+}
+
+func (r *rsaPrivateKey) encryptRSA(data []byte) ([]byte, error) {
+	hash := r.getHash()
+	cParam := (C.CK_RSA_PKCS_OAEP_PARAMS_PTR)(C.malloc(C.sizeof_CK_RSA_PKCS_OAEP_PARAMS))
+	defer C.free(unsafe.Pointer(cParam))
+
+	switch *hash {
+	case crypto.SHA256:
+		cParam.hashAlg = C.CKM_SHA256
+		cParam.mgf = C.CKG_MGF1_SHA256
+	case crypto.SHA384:
+		cParam.hashAlg = C.CKM_SHA384
+		cParam.mgf = C.CKG_MGF1_SHA384
+	case crypto.SHA512:
+		cParam.hashAlg = C.CKM_SHA512
+		cParam.mgf = C.CKG_MGF1_SHA512
+	case crypto.SHA1:
+		cParam.hashAlg = C.CKM_SHA_1
+		cParam.mgf = C.CKG_MGF1_SHA1
+	default:
+		return nil, fmt.Errorf("unsupported hash algorithm: %s", hash)
+	}
+
+	cDataBytes := make([]C.CK_BYTE, len(data))
+	for i, b := range cDataBytes {
+		cDataBytes[i] = C.CK_BYTE(b)
+	}
+
+	cParam.source = C.CKZ_DATA_SPECIFIED
+	cParam.pSourceData = nil
+	cParam.ulSourceDataLen = 0
+
+	m := C.CK_MECHANISM{
+		mechanism:      C.CKM_RSA_PKCS_OAEP,
+		pParameter:     C.CK_VOID_PTR(cParam),
+		ulParameterLen: C.CK_ULONG(C.sizeof_CK_RSA_PKCS_OAEP_PARAMS),
+	}
+
+	rv := C.ck_encrypt_init(r.o.fl, r.o.h, &m, r.pubH.o)
+	if err := isOk("C_EncryptInit", rv); err != nil {
+		return nil, err
+	}
+
+	// RSA is only able to encrypt data to a maximum amount equal to your key size
```

Comment:
You're API should encrypt the entire buffer. I also do not believe this comment is true.

Can you point me to here this is stated?

Reply:
I found information on RSA maximum encryption size here: https://mbed-tls.readthedocs.io/en/latest/kb/cryptography/rsa-encryption-maximum-data-size/#:~:text=RSA%20is%20only%20able%20to,is%20not%20designed%20for%20this).

Reply:
I will update this to follow the specification in 5.2 instead of just using the key size